pkg/secretpath/path.go

```diff
@@ -0,0 +1,226 @@
+// Package secretpath implements utility functions for manipulating paths
+// compatible with SecretHub, e.g. namespaces, repositories, directories,
+// secrets and versions.
+package secretpath
+
+import (
+	"strconv"
+	"strings"
+
+	"bitbucket.org/zombiezen/cardcpx/natsort"
+)
+
+const (
+	elemSepByte    = '/'
+	elemSep        = string(elemSepByte)
+	versionSepByte = ':'
+	versionSep     = string(versionSepByte)
+	latestSuffix   = ":latest"
+)
+
+// Join joins any number of elements into a path, adding a separator
+// if necessary. Empty string elements are ignored.
+func Join(elements ...string) string {
+	result := ""
+	for _, e := range elements {
+		e = strings.Trim(e, elemSep)
+		if len(e) != 0 && len(result) == 0 {
+			result += e
+		} else if len(e) != 0 {
+			result += elemSep + e
+		}
+	}
+
+	return Clean(result)
+}
+
+// HasVersion returns true when a version suffix is specified in the path.
+func HasVersion(path string) bool {
+	return hasVersion(path)
+}
+
+// hasVersion returns true when a version suffix is specified in the path.
+func hasVersion(path string) bool {
+	i := strings.LastIndex(path, versionSep)
+	return i >= 0 && i < len(path)-1
+}
+
+// Version returns the version number suffix of a path, returning -1
+// when :latest or negative version numbers are given. If no version
+// suffix is set or it cannot be parsed to an integer, 0 is returned.
+func Version(path string) int {
```

Comment:
Should we use the same return for valid and invalid inputs?

We now have valid inputs:
`:latest` => -1
`` => 0

That return the same as these invalid inputs:
`:foo` => 0
`:-1` => -1

Reply:
`-1` is a valid(ish) input, but in regards to your question: I think it should ignore the version when it's empty and it should ignore the version when it's not valid. So that would be the same case.

Reply:
> I think it should ignore the version when it's empty and it should ignore the version when it's not valid

What is `it` in this case?

---

If applications do not error on invalid versions, do invalid versions become valid input?

---

There are applications that want to treat empty versions and invalid versions differently. For example, a linter for run env files of the CLI will want to error on invalid versions, while accepting empty versions.

---

> -1 is a valid(ish) input

What do you mean with this? Should it be possible to supply `-1` as secret path version?  It's currently not accepted by `api.ValidateSecretPath`, so is not accepted by any of our applications (including the server).